src/coord/src/coord/dataflow_builder.rs

```diff
@@ -213,11 +235,164 @@ impl<'a> DataflowBuilder<'a> {
+                session,
+            } => {
+                expr.visit_mut_post(&mut |e| {
+                    if let MirScalarExpr::CallNullary(f) = e {
+                        *e = self.eval_nullary_func(f, logical_time, session);
+                    }
+                });
+            }
+
+            // Reject the query if it contains any nullary function calls.
+            ExprPrepStyle::Index => {
+                let mut last_observed_nullary_func = None;
+                expr.visit_mut_post(&mut |e| {
+                    if let MirScalarExpr::CallNullary(f) = e {
+                        last_observed_nullary_func = Some(f.clone());
+                    }
+                });
+                if let Some(f) = last_observed_nullary_func {
+                    return Err(CoordError::UnmaterializableFunction(f));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn eval_nullary_func(
+        &self,
+        f: &NullaryFunc,
+        logical_time: u64,
+        session: &Session,
+    ) -> MirScalarExpr {
+        let pack_1d_array = |datums: Vec<Datum>| {
+            let mut row = Row::default();
+            row.push_array(
+                &[ArrayDimension {
+                    lower_bound: 1,
+                    length: datums.len(),
+                }],
+                datums,
+            )
+            .expect("known to be a valid array");
+            MirScalarExpr::Literal(Ok(row), f.output_type())
+        };
+        let pack = |datum| MirScalarExpr::literal_ok(datum, f.output_type().scalar_type);
+
+        match f {
+            NullaryFunc::CurrentDatabase => pack(Datum::from(session.vars().database())),
+            NullaryFunc::CurrentSchemasWithSystem => pack_1d_array(
+                session
+                    .vars()
+                    .search_path()
+                    .iter()
+                    .map(|s| Datum::String(s))
+                    .collect(),
+            ),
+            NullaryFunc::CurrentSchemasWithoutSystem => {
+                use crate::catalog::builtin::{
+                    INFORMATION_SCHEMA, MZ_CATALOG_SCHEMA, MZ_INTERNAL_SCHEMA, MZ_TEMP_SCHEMA,
+                    PG_CATALOG_SCHEMA,
+                };
+                pack_1d_array(
+                    session
+                        .vars()
+                        .search_path()
+                        .iter()
+                        .filter(|s| {
+                            (**s != PG_CATALOG_SCHEMA)
+                                && (**s != INFORMATION_SCHEMA)
+                                && (**s != MZ_CATALOG_SCHEMA)
+                                && (**s != MZ_TEMP_SCHEMA)
+                                && (**s != MZ_INTERNAL_SCHEMA)
+                        })
+                        .map(|s| Datum::String(s))
+                        .collect(),
+                )
+            }
+            NullaryFunc::CurrentTimestamp => pack(Datum::from(session.pcx().wall_time)),
```

Comment:
uh can we remove `pcx` from `StatementContext` now? Can we also remove `catalog` from `StatementContext`?!

Reply:
Almost! I was hoping I might be able to leave that to you in a followup. I think getting rid of `pcx` is pretty straightforward, but I didn't want to futz with the way transactions install `pcx` in the `Session` since this PR was already getting pretty large. Getting rid of `catalog` is harder, because it's 100% needed for name resolution... but that's exactly the work that Justin had started on way back when! If we hoisted the name resolution pass out of planning, then at least planning would proceed without the catalog. There are also like three things in the `CREATE SOURCE`/`CREATE SINK` pathway that depend on the catalog that could be fixed to not, but would require annoying plumbing to rejigger.

Reply:
:+1: